### DISCUSSION TITLE
postgre database cartridge - minimum install options (do I need RDK_PGSQL_STATIC)

### DISCUSSION THREAD

**Discussion by shortydutchie** [rdkit/rdkit#5158] [Q&A]
I am struggling quite a bit with the database cartridge installation...

I have a working globally installed postgre serving various applications (linux Ubuntu 20.04) - I'd like to  keep that and install the cartridge on some of its databases. (*)

(1)
When I install the cartridge from the suggested conda environment approach in the docs, what are the minimum requirements I need to specify for a build with 'make'? 
*  RDK_BUILD_PGSQL=ON 
*  Do I need RDK_PGSQL_STATIC=ON as well?

(2)
Should 'sudo make install' work for installing into my system's postgre installation? 
If I do not sudo, where does it end up (and how to tell PG where the cartridge is?)

(3)
Is the cartridge installation self-contained, or do I need to make sure that any rdkit I use within my app in a conda shell is exactly the same as the one used to build the cartridge?

Any pointers appreciated, any misunderstandings are due to my own stupidity...

**Update 220405 -** 

Various setups tried, with some progress -
* You should not create a 'build' dir and cmake there as per the docs, as the FindNumPy.cmake override for numpy is not picked up then.
* cmake needs to be a later version than I had (I upgraded to 3.22.1 which result in fewer errors)
* You need to : sudo apt-get install libpq-dev
* Need to set -DBOOST_ROOT="$CONDA_PREFIX" (boost libs are installed in the virtual env)
* addition of -DRDK_BUILD_PGSQL=ON should build the cartridge

This fails and the issue seems to be in cmake; CMakeLists.txt invokes:

   **find_package(PostgreSQL REQUIRED)**

and this does not seem to work for my version of postgresql, 12.9. Above cannot derive the relevant paths variables from the installation for some reason it seems. 

I tried addition of 
   -DPostgreSQL_LIBRARY=/usr/lib/postgresql/12/ (which contains bin and lib subdirs); 
and/or 
   -DPostgreSQL_ROOT=/usr/lib/postgresql/12/

The variable PostgreSQL_CONFIG is then correctly set to 
    /usr/lib/postgresql/12/bin/pg_config 
with above setting (previously it set a wrong path).

This still results in a warning though, which leads to an error later on (prob because something is not set correctly here):

    CMake Warning (dev) at Code/PgSQL/rdkit/CMakeLists.txt:23 (find_package):
     Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.

So this is not working.

There is no override .cmake file (FindPostGreSQL.cmake?) in the Modules directory (Code/cmake/Modules/), seems like one is required. Does anyone have one that works with postgresql 12 (12.9) ?
(I think postgresql is now at version 14...)

Or do I need to use an older cmake perhaps? (current is 3.22.1)

Any pointers appreciated
Thanks
Will

(*)
Running a second parallel postgre instance in a conda environment gave me too much grief and seems wasteful.
I don't want to go the docker way at this point & rebuild databases there.

**Comment by ptosco** [comment]
@shortydutchie Hi Will, these are the `cmake` settings that I currently use to build against Postgres 13:
```
-D RDK_BUILD_PGSQL=ON \
-D RDK_PGSQL_STATIC=ON \
-D PostgreSQL_INCLUDE_DIR=/install/postgres_13_1/include \
-D PostgreSQL_TYPE_INCLUDE_DIR=/install/postgres_13_1/include/server \
-D PostgreSQL_LIBRARY=/install/postgres_13_1/lib/libpq.so \
```
I do not set `PostgreSQL_ROOT`.
You might need to change my paths slightly as the details of PostgreSQL directory locations change across distributions, but other than that it should work.

**Reply by shortydutchie** [reply]
Thank you, I'll see if l can upgrade pg to 13 and give it another go. Maybe it has a slightly different directory structure. I did try some variations that were unsuccessful so far.

Which version of cmake do you use?

Much appreciated,
Will

**Reply by ptosco** [reply]
I don’t think the Postgres version will make a difference. I’ll take a look at my WSL install which is based on Ubuntu 20.04 and get back to you.

**Comment by ptosco** [comment]
@shortydutchie These are the settings I have been using with success on WSL/Ubuntu 20.04:
```    -D RDK_BUILD_PGSQL=ON \
    -D RDK_PGSQL_STATIC=OFF \
    -D PostgreSQL_CONFIG_DIR=/usr/lib/postgresql/14/bin \
    -D PostgreSQL_INCLUDE_DIR="/usr/include/postgresql" \
    -D PostgreSQL_TYPE_INCLUDE_DIR="/usr/include/postgresql/14/server" \
    -D PostgreSQL_LIBRARY="/usr/lib/x86_64-linux-gnu/libpq.so" \
```
